PR: https://github.com/odoo/odoo/pull/212679

From: 2713876dbc70d3984e584a9037a2206dcda4e84a
From: Arnold Moyaux
Date: 2025-09-07 15:50:27

Breaking data model changes scores: del:5 + add:2, change matches:
-    group_id = fields.Many2one('procurement.group', string="Procurement Group", copy=False)
+    reference_ids = fields.Many2many(
-    group_id = fields.Many2one('procurement.group', 'Procurement group that generated this line', index='btree_not_null')
-    _inherit = 'procurement.group'
-    purchase_line_ids = fields.One2many('purchase.order.line', 'group_id', string='Linked Purchase Order Lines', copy=False)
+    purchase_ids = fields.Many2many(

Total Changes: 162

[REF] stock, *: add reference

The model `procurement.group` is replaced by a more flexible one:
`stock.reference`. The first one was somehow not well understood and a
bit limited by its many2one links to other business model (Sales Order,
Purchase Order, ...). The reference is made of a name and many2many to
all the documents integrated with Inventory. It's purpose has been
shrank to the minimal "make links between object". For instance
opening the manufacturing order created for a MTO sale. The merge of
stock move is delegated to existing parameters (`sale_id`,
created_purchase_line_ids`, ...)

The `mrp.production` needed an extra effort to group them according to
inheritance relationships. This commit add a new object to group
productions together either in a backorder relation or in a child-parent
one. This allow for instance to have a Sale Order linked to the MTO
product of the sold product and the production them selves to be linked
to the SO *and* the potential child MOs.

The procurement group propagation, before set on the Stock Rule, allowing
the stock moves to be merged in an existing picking or not has been
removed and replaced (for the creation of purchase order only) by a new
parameter on `res.partner` allowing the group of Request for Quotation
base on the scheduled date of the potential merge candidate (never, same
week, same month, always). All other merge scenarios should use batches
and waves transfers.

closes odoo/odoo#212679

Task: 4279803
Related: odoo/enterprise#87170
Related: odoo/upgrade#8326
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/purchase_stock/models/__init__.py
+++ b/addons/purchase_stock/models/__init__.py
@@ -12,4 +12,5 @@ from . import res_partner
 from . import res_company
 from . import stock
 from . import stock_move
+from . import stock_reference
 from . import stock_rule

--- a/addons/purchase_stock/models/purchase_order.py
+++ b/addons/purchase_stock/models/purchase_order.py
@@ -24,7 +24,9 @@ class PurchaseOrder(models.Model):
         help="This will determine operation type of incoming shipment")
     default_location_dest_id_usage = fields.Selection(related='picking_type_id.default_location_dest_id.usage', string='Destination Location Type',
         help="Technical field used to display the Drop Ship Address", readonly=True)
-    group_id = fields.Many2one('procurement.group', string="Procurement Group", copy=False)
+    reference_ids = fields.Many2many(
+        'stock.reference', 'stock_reference_purchase_rel', 'purchase_id',
+        'reference_id', string='References', copy=False)
     is_shipped = fields.Boolean(compute="_compute_is_shipped")
     effective_date = fields.Datetime("Arrival", compute='_compute_effective_date', store=True, copy=False,
         help="Completion date of the first receipt order.")
@@ -186,6 +188,9 @@ class PurchaseOrder(models.Model):
                 if picking.state == 'done':
                     picking.message_post(body=self.env._("The purchase order %s this receipt is linked to was cancelled.", order._get_html_link()))
 
+            if order.reference_ids:
+                order.reference_ids.purchase_ids = [Command.unlink(order.id)]
+
         order_lines = self.env['purchase.order.line'].browse(order_lines_ids)
         moves_to_cancel_ids = OrderedSet()
         moves_to_recompute_ids = OrderedSet()
@@ -204,8 +209,6 @@ class PurchaseOrder(models.Model):
                     moves_to_cancel_ids.update(move_dest_ids.ids)
                 else:
                     moves_to_recompute_ids.update(move_dest_ids.ids)
-            if order_line.group_id:
-                order_line.group_id.purchase_line_ids = [Command.unlink(order_line.id)]
 
         if moves_to_cancel_ids:
             moves_to_cancel = self.env['stock.move'].browse(moves_to_cancel_ids)
@@ -332,16 +335,15 @@ class PurchaseOrder(models.Model):
             picking_type = self.env['stock.picking.type'].with_context(active_test=False).search([('code', '=', 'incoming'), ('warehouse_id', '=', False)])
         return picking_type[:1]
 
-    def _prepare_group_vals(self):
+    def _prepare_reference_vals(self):
         self.ensure_one()
         return {
             'name': self.name,
-            'partner_id': self.partner_id.id,
         }
 
     def _prepare_picking(self):
-        if not self.group_id:
-            self.group_id = self.group_id.create(self._prepare_group_vals())
+        if not self.reference_ids:
+            self.reference_ids = self.reference_ids.create(self._prepare_reference_vals())
         if not self.partner_id.property_stock_supplier.id:
             raise UserError(_("You must set a Vendor Location for this partner %s", self.partner_id.name))
         return {
@@ -353,6 +355,7 @@ class PurchaseOrder(models.Model):
             'location_id': self.partner_id.property_stock_supplier.id,
             'company_id': self.company_id.id,
             'state': 'draft',
+            'reference_ids': [Command.set(self.reference_ids.ids)],
         }
 
     def _create_picking(self):

--- a/addons/purchase_stock/models/purchase_order_line.py
+++ b/addons/purchase_stock/models/purchase_order_line.py
@@ -1,5 +1,7 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
+from dateutil.relativedelta import relativedelta
+
 from odoo import SUPERUSER_ID, api, Command, fields, models, _
 from odoo.tools.float_utils import float_compare, float_is_zero, float_round
 from odoo.exceptions import UserError
@@ -24,14 +26,13 @@ class PurchaseOrderLine(models.Model):
                                            ondelete={'stock_moves': _ondelete_stock_moves})
 
     move_ids = fields.One2many('stock.move', 'purchase_line_id', string='Reservation', readonly=True, copy=False)
-    orderpoint_id = fields.Many2one('stock.warehouse.orderpoint', 'Orderpoint', copy=False, index='btree_not_null')
+    orderpoint_id = fields.Many2one('stock.warehouse.orderpoint', 'Orderpoint', copy=False, index='btree_not_null', ondelete='set null')
     move_dest_ids = fields.Many2many('stock.move', 'stock_move_created_purchase_line_rel', 'created_purchase_line_id', 'move_id', 'Downstream moves alt')
     product_description_variants = fields.Char('Custom Description')
     propagate_cancel = fields.Boolean('Propagate cancellation', default=True)
     forecasted_issue = fields.Boolean(compute='_compute_forecasted_issue')
     is_storable = fields.Boolean(related='product_id.is_storable')
     location_final_id = fields.Many2one('stock.location', 'Location from procurement')
-    group_id = fields.Many2one('procurement.group', 'Procurement group that generated this line', index='btree_not_null')
 
     def _compute_qty_received_method(self):
         super(PurchaseOrderLine, self)._compute_qty_received_method()
@@ -296,7 +297,7 @@ class PurchaseOrderLine(models.Model):
             'company_id': self.order_id.company_id.id,
             'price_unit': price_unit,
             'picking_type_id': self.order_id.picking_type_id.id,
-            'group_id': self.order_id.group_id.id,
+            'reference_ids': [Command.set(self.order_id.reference_ids.ids)],
             'origin': self.order_id.name,
             'propagate_cancel': self.propagate_cancel,
             'warehouse_id': self.order_id.picking_type_id.warehouse_id.id,
@@ -337,17 +338,16 @@ class PurchaseOrderLine(models.Model):
         if line_description and product_id.name != line_description:
             res['name'] = (res['name'] + '\n' + line_description).strip()
         res['date_planned'] = values.get('date_planned')
+        # The date must be day before or equal at the supplier target day
+        if po.partner_id.group_rfq == 'week' and po.partner_id.group_on != 'default':
+            delta_days = (res['date_planned'].isoweekday() - int(po.partner_id.group_on)) % 7
+            res['date_planned'] = fields.Datetime.to_datetime(res['date_planned']) - relativedelta(days=delta_days)
         res['move_dest_ids'] = [(4, x.id) for x in values.get('move_dest_ids', [])]
         res['location_final_id'] = location_dest_id.id
         res['orderpoint_id'] = values.get('orderpoint_id', False) and values.get('orderpoint_id').id
         res['propagate_cancel'] = values.get('propagate_cancel')
         res['product_description_variants'] = values.get('product_description_variants')
         res['product_no_variant_attribute_value_ids'] = values.get('never_product_template_attribute_value_ids')
-
-        # Need to attach purchase order to procurement group for mtso
-        group = values.get('group_id')
-        if group and not res['move_dest_ids']:
-            res['group_id'] = values['group_id'].id
         return res
 
     def _create_stock_moves(self, picking):
@@ -355,7 +355,6 @@ class PurchaseOrderLine(models.Model):
         for line in self.filtered(lambda l: not l.display_type):
             for val in line._prepare_stock_moves(picking):
                 values.append(val)
-            line.move_dest_ids.created_purchase_line_ids = [Command.clear()]
 
         return self.env['stock.move'].create(values)
 
@@ -369,7 +368,7 @@ class PurchaseOrderLine(models.Model):
             description_picking = values['product_description_variants']
         lines = self.filtered(
             lambda l: l.propagate_cancel == values['propagate_cancel']
-            and (l.orderpoint_id == values['orderpoint_id'] if values['orderpoint_id'] and not values['move_dest_ids'] else True)
+            and (l.orderpoint_id in [values['orderpoint_id'], False] if values['orderpoint_id'] and not values['move_dest_ids'] else True)
             and (l.product_uom_id == product_uom if values.get('force_uom') else True)
         )
 
@@ -386,10 +385,7 @@ class PurchaseOrderLine(models.Model):
             if product_lang.description_purchase:
                 name += '\n' + product_lang.description_purchase
             lines = lines.filtered(lambda l: (l.name == name + '\n' + description_picking) or (values.get('product_description_variants') in (product_lang.name, product_id.with_user(SUPERUSER_ID).name) and l.name == name))
-            if lines:
-                return lines[0]
-
-        return lines and lines[0] or self.env['purchase.order.line']
+        return lines and lines.sorted(lambda l: l.orderpoint_id)[0] or self.env['purchase.order.line']
 
     def _get_outgoing_incoming_moves(self):
         outgoing_moves = self.env['stock.move']

--- a/addons/purchase_stock/models/res_partner.py
+++ b/addons/purchase_stock/models/res_partner.py
@@ -18,6 +18,24 @@ class ResPartner(models.Model):
     suggest_based_on = fields.Char(store=True, default='30_days')
     suggest_days = fields.Integer(store=True, default=7)
     suggest_percent = fields.Integer(store=True, default=100)
+    group_rfq = fields.Selection(
+        [('default', 'On Order'), ('day', 'Daily'), ('week', 'Weekly'), ('all', 'Always')],
+        string='Group RFQ', required=True, default='default', help="Define if RFQ should be grouped \
+        together based on expected arrival.\n \
+        On Order: Replenishment needs will be grouped together except for MTO.\n \
+        Daily: Replenishment needs will be grouped if the expected arrival is the same day\n \
+        Weekly: Replenishment needs will be grouped if the expected arrival is the same week or week day\n \
+        Always: Replenishment needs will always be grouped.")
+    group_on = fields.Selection([
+        ('default', 'Expected Date'),
+        ('1', 'Monday'),
+        ('2', 'Tuesday'),
+        ('3', 'Wednesday'),
+        ('4', 'Thursday'),
+        ('5', 'Friday'),
+        ('6', 'Saturday'),
+        ('7', 'Sunday'),
+    ], string='Week Day', required=True, default='default')
 
     @api.depends('purchase_line_ids')
     def _compute_on_time_rate(self):

--- a/addons/purchase_stock/models/stock.py
+++ b/addons/purchase_stock/models/stock.py
@@ -54,7 +54,6 @@ class StockWarehouse(models.Model):
                 'create_values': {
                     'action': 'buy',
                     'picking_type_id': self.in_type_id.id,
-                    'group_propagation_option': 'none',
                     'company_id': self.company_id.id,
                     'route_id': self._find_or_create_global_route('purchase_stock.route_warehouse0_buy', _('Buy')).id,
                     'propagate_cancel': self.reception_steps != 'one_step',
@@ -268,8 +267,8 @@ class StockWarehouseOrderpoint(models.Model):
             }
         return super()._get_replenishment_order_notification()
 
-    def _prepare_procurement_values(self, date=False, group=False):
-        values = super()._prepare_procurement_values(date=date, group=group)
+    def _prepare_procurement_values(self, date=False):
+        values = super()._prepare_procurement_values(date=date)
         values['supplierinfo_id'] = self.supplier_id
         return values
 
@@ -324,22 +323,3 @@ class StockLot(models.Model):
         action['domain'] = [('id', 'in', self.mapped('purchase_order_ids.id'))]
         action['context'] = dict(self.env.context, create=False)
         return action
-
-
-class ProcurementGroup(models.Model):
-    _inherit = 'procurement.group'
-
-    purchase_line_ids = fields.One2many('purchase.order.line', 'group_id', string='Linked Purchase Order Lines', copy=False)
-
-    @api.model
-    def run(self, procurements, raise_user_error=True):
-        wh_by_comp = dict()
-        for procurement in procurements:
-            routes = procurement.values.get('route_ids')
-            if routes and any(r.action == 'buy' for r in routes.rule_ids):
-                company = procurement.company_id
-                if company not in wh_by_comp:
-                    wh_by_comp[company] = self.env['stock.warehouse'].search([('company_id', '=', company.id)])
-                wh = wh_by_comp[company]
-                procurement.values['route_ids'] |= wh.reception_route_id
-        return super().run(procurements, raise_user_error=raise_user_error)

--- a/addons/purchase_stock/models/stock_move.py
+++ b/addons/purchase_stock/models/stock_move.py
@@ -25,10 +25,7 @@ class StockMove(models.Model):
 
     @api.model
     def _prepare_merge_negative_moves_excluded_distinct_fields(self):
-        excluded_fields = super()._prepare_merge_negative_moves_excluded_distinct_fields() + ['created_purchase_line_ids']
-        if self.env['ir.config_parameter'].sudo().get_param('purchase_stock.merge_different_procurement'):
-            excluded_fields += ['procure_method']
-        return excluded_fields
+        return super()._prepare_merge_negative_moves_excluded_distinct_fields() + ['created_purchase_line_ids']
 
     @api.depends('purchase_line_id', 'purchase_line_id.product_uom_id')
     def _compute_packaging_uom_id(self):
@@ -94,12 +91,17 @@ class StockMove(models.Model):
         self.ensure_one()
         return self.origin_returned_move_id or not self.purchase_line_id or not self.product_id.id
 
+    def _prepare_extra_move_vals(self, qty):
+        vals = super()._prepare_extra_move_vals(qty)
+        vals['purchase_line_id'] = self.purchase_line_id.id
+        return vals
+
     def _prepare_move_split_vals(self, uom_qty):
         vals = super(StockMove, self)._prepare_move_split_vals(uom_qty)
-        vals['purchase_line_id'] = self.purchase_line_id.id
         # when backordering an mto move link the bakcorder to the purchase order
         if self.procure_method == 'make_to_order' and self.created_purchase_line_ids:
             vals['created_purchase_line_ids'] = [Command.set(self.created_purchase_line_ids.ids)]
+        vals['purchase_line_id'] = self.purchase_line_id.id
         return vals
 
     def _clean_merged(self):

--- a/None
+++ b/addons/purchase_stock/models/stock_reference.py
@@ -0,0 +1,9 @@
+from odoo import fields, models
+
+
+class StockReference(models.Model):
+    _inherit = 'stock.reference'
+
+    purchase_ids = fields.Many2many(
+        'purchase.order', 'stock_reference_purchase_rel', 'reference_id',
+        'purchase_id', string="Purchases", copy=False)

--- a/addons/purchase_stock/models/stock_rule.py
+++ b/addons/purchase_stock/models/stock_rule.py
@@ -7,7 +7,7 @@ from dateutil.relativedelta import relativedelta
 from markupsafe import Markup
 from odoo.tools import float_compare
 
-from odoo import api, fields, models, SUPERUSER_ID, _
+from odoo import api, fields, models, SUPERUSER_ID, _, Command
 from odoo.addons.stock.models.stock_rule import ProcurementException
 from odoo.tools import groupby
 
@@ -42,6 +42,19 @@ class StockRule(models.Model):
         if self.action == 'buy':
             self.location_src_id = False
 
+    @api.model
+    def run(self, procurements, raise_user_error=True):
+        wh_by_comp = dict()
+        for procurement in procurements:
+            routes = procurement.values.get('route_ids')
+            if routes and any(r.action == 'buy' for r in routes.rule_ids):
+                company = procurement.company_id
+                if company not in wh_by_comp:
+                    wh_by_comp[company] = self.env['stock.warehouse'].search([('company_id', '=', company.id)])
+                wh = wh_by_comp[company]
+                procurement.values['route_ids'] |= wh.reception_route_id
+        return super().run(procurements, raise_user_error=raise_user_error)
+
     @api.model
     def _run_buy(self, procurements):
         procurements_by_po_domain = defaultdict(list)
@@ -101,7 +114,11 @@ class StockRule(models.Model):
                     # Indeed, the current user may be a user without access to Purchase, or even be a portal user.
                     po = self.env['purchase.order'].with_company(company_id).with_user(SUPERUSER_ID).create(vals)
             else:
+                reference_ids = set()
+                for procurement in procurements:
+                    reference_ids |= set(procurement.values.get('reference_ids', self.env['stock.reference']).ids)
                 # If a purchase order is found, adapt its `origin` field.
+                po.reference_ids = [Command.link(ref_id) for ref_id in reference_ids]
                 if po.origin:
                     missing_origins = origins - set(po.origin.split(', '))
                     if missing_origins:
@@ -255,7 +272,7 @@ class StockRule(models.Model):
                 'move_dest_ids': move_dest_ids,
                 'orderpoint_id': orderpoint_id,
             })
-            merged_procurement = self.env['procurement.group'].Procurement(
+            merged_procurement = self.env['stock.rule'].Procurement(
                 procurement.product_id, quantity, procurement.product_uom,
                 procurement.location_id, procurement.name, procurement.origin,
                 procurement.company_id, values
@@ -309,10 +326,6 @@ class StockRule(models.Model):
 
         fpos = self.env['account.fiscal.position'].with_company(company_id)._get_fiscal_position(partner)
 
-        gpo = self.group_propagation_option
-        group = (gpo == 'fixed' and self.group_id.id) or \
-                (gpo == 'propagate' and values.get('group_id') and values['group_id'].id) or False
-
         return {
             'partner_id': partner.id,
             'user_id': partner.buyer_id.id,
@@ -324,17 +337,13 @@ class StockRule(models.Model):
             'payment_term_id': partner.with_company(company_id).property_supplier_payment_term_id.id,
             'date_order': purchase_date,
             'fiscal_position_id': fpos.id,
-            'group_id': group
+            'reference_ids': [Command.set(values.get('reference_ids', self.env['stock.reference']).ids)],
         }
 
     def _make_po_get_domain(self, company_id, values, partner):
-        gpo = self.group_propagation_option
-        group = (gpo == 'fixed' and self.group_id) or \
-                (gpo == 'propagate' and 'group_id' in values and values['group_id']) or False
         currency = ('supplier' in values and values['supplier'].currency_id) or \
                    partner.with_company(company_id).property_purchase_currency_id or \
                    company_id.currency_id
-
         domain = (
             ('partner_id', '=', partner.id),
             ('state', '=', 'draft'),
@@ -343,16 +352,20 @@ class StockRule(models.Model):
             ('user_id', '=', partner.buyer_id.id),
             ('currency_id', '=', currency.id),
         )
-        delta_days = self.env['ir.config_parameter'].sudo().get_param('purchase_stock.delta_days_merge')
-        if values.get('orderpoint_id') and delta_days is not False:
-            procurement_date = fields.Date.to_date(values['date_planned']) - relativedelta(days=int(values['supplier'].delay))
-            delta_days = int(delta_days)
-            domain += (
-                ('date_order', '<=', datetime.combine(procurement_date + relativedelta(days=delta_days), datetime.max.time())),
-                ('date_order', '>=', datetime.combine(procurement_date - relativedelta(days=delta_days), datetime.min.time()))
-            )
-        if group:
-            domain += (('group_id', '=', group.id),)
+        if partner.group_rfq == 'default':
+            if values.get('reference_ids'):
+                domain += (('reference_ids', 'in', tuple(values['reference_ids'].ids)),)
+            else:
+                domain += (('reference_ids', '=', False),)
+        date_planned = fields.Datetime.from_string(values['date_planned']) - relativedelta(days=int(values['supplier'].delay))
+        if partner.group_rfq == 'day':
+            start_dt = datetime.combine(date_planned, datetime.min.time())
+            end_dt = datetime.combine(date_planned, datetime.max.time())
+            domain += (('date_planned', '>=', start_dt), ('date_planned', '<=', end_dt))
+        if partner.group_rfq == 'week':
+            start_dt = datetime.combine(date_planned - relativedelta(days=date_planned.weekday()), datetime.min.time())
+            end_dt = datetime.combine(date_planned + relativedelta(days=6 - date_planned.weekday()), datetime.max.time())
+            domain += (('date_planned', '>=', start_dt), ('date_planned', '<=', end_dt))
         return domain
 
     def _push_prepare_move_copy_values(self, move_to_copy, new_date):
@@ -363,4 +376,4 @@ class StockRule(models.Model):
         return res
 
     def _get_partner_id(self, values, rule):
-        return values.get("supplierinfo_name") or (values.get("force_uom") and values.get("group_id") and values.get("group_id").partner_id)
+        return values.get("supplierinfo_name") or (values.get("force_uom") and values.get("partner"))
